Ruby Course (various): Link updates as per Issue 28290 (#28392)
TheOdinProject/curriculum: ruby/testing_ruby_with_rspec/test_driven_development.md

### What is test-driven development?

> As outlined in [Wikipedia](https://en.wikipedia.org/wiki/Test-driven_development), Test-Driven Development (TDD) is a process and technique of software development that relies on the repetition of a very short and specific development cycle. In each cycle, requirements (i.e., what you want your code to do) are turned into specific test cases first. These requirements could be anything from an entire feature that requires end-to-end (E2E) testing, such as tests that cover a user logging into your website successfully AND unsuccessfully, to a new Ruby class you've devised, for which unit tests might suffice. Either way, the test suite for these requirements fail initially, since actual code hasn't been written yet.

Reviewer: Can this be reverted back to the original? Same reasons as above regarding the new link text, and the original is sufficiently accessible.

Once the code is written and passes our test suite, you can move onto refactoring. As you might have guessed, running the test suite against the changes we make to our features or units enables us to refactor *while* making sure everything still works.

Colloquially, this process is often referred to as the "**red-green-refactor**" cycle. That's all it is, in a nutshell: automated tests **drive** the design of software. You don't need to know how the entire architecture of your sweet, new, industry-breaking application will work. Your application only has to be broken down, step-by-step, until small units are identified and covered by tests.

### Why is it important?

Even if TDD intuitively seems like good practice, its utility is hotly debated even today.

> No one is saying that tests aren't useful. Testing at all levels is *crucial*. Joël Quenneville's article on the Rails [Testing Pyramid](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid) provides an overview into the different types and levels of testing that Rails developers routinely use.

Reviewer: Good catch! Missed that from my spreadsheet :pray:

1. James Sinclair's article explains more about [the why/how and value behind TDD](https://jrsinclair.com/articles/2016/one-weird-trick-that-will-change-the-way-you-code-forever-javascript-tdd/) and also includes some great examples of how to apply it. While the examples are in JavaScript, the focus is on the value behind the TDD approach.

</div>

### Knowledge check

The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.

- [What is the "red-green-refactor" cycle?](#what-is-test-driven-development)
- [What is the difference between unit and integration tests?](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid)
- [How does TDD help in code refactoring?](https://www.madetech.com/blog/9-benefits-of-test-driven-development)

### Additional resources

This section contains helpful links to related content. It isn't required, so consider it supplemental.

> - Youtube has [a video](https://www.youtube.com/watch?v=PCEHRFHKZSk) that provides counter arguments for DHH’s [“TDD is Dead. Long Live Testing”](https://dhh.dk/2014/tdd-is-dead-long-live-testing.html).

Reviewer: Same reasons as earlier
```suggestion
- This video that provides [counter arguments for DHH's “TDD is Dead. Long Live Testing” article](https://www.youtube.com/watch?v=PCEHRFHKZSk) from the assignment.
```

> - Martin Fowler has [a series of talks](https://martinfowler.com/articles/is-tdd-dead/) that provides an even more rounded take on TDD.

Reviewer: ```suggestion
- For a more rounded taken on TDD, watch or listen to this [series of talks on the pros and cons of TDD](https://martinfowler.com/articles/is-tdd-dead/).
```
